conformance/src/main/scala/za/co/absa/enceladus/conformance/interpreter/OptimizerTimeTracker.scala

```diff
@@ -0,0 +1,126 @@
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.enceladus.conformance.interpreter
+
+import org.apache.log4j.LogManager
+import org.apache.spark.sql.{DataFrame, SparkSession}
+import org.apache.spark.sql.functions.{col, lit}
+import za.co.absa.enceladus.utils.general.JsonUtils
+import za.co.absa.enceladus.utils.schema.SchemaUtils
+
+class OptimizerTimeTracker(inputDf: DataFrame, isWorkaroundEnabled: Boolean)(implicit spark: SparkSession) {
+  private val log = LogManager.getLogger(this.getClass)
+
+  private val maxToleratedPlanGenerationPerRuleMs = 100L
+  private val initialElapsedTimeBaselineMs = 300L
+  private var baselineTimeMs = initialElapsedTimeBaselineMs
+  private var lastExecutionPlanOptimizatioTime = 0L
+
+  private val idField1 = SchemaUtils.getUniqueName("tmpId", Option(inputDf.schema))
+  private val idField2 = s"${idField1}_2"
+  private val dfWithId = inputDf.withColumn(idField1, lit(1))
+  private val dfJustId = JsonUtils.getDataFrameFromJson(spark, Seq(s"""{"$idField2":1}""")).cache
+
+  /**
+    * Returns a dataframe prepared to apply the Catalyst workaround
+    */
+  def getWorkaroundDataframe: DataFrame = {
+    if (isWorkaroundEnabled) {
+      dfWithId
+    } else {
+      inputDf
+    }
+  }
+
+  /**
+    * Cleans up the additional field used for applying the Catalyst workaround
+    */
+  def cleanupWorkaroundDf(df: DataFrame): DataFrame = df.drop(col(idField1))
+
+  /**
+    * Returns true of a dataframe might require a Catalyst issue workaround.
+    * A workaround is required if execution plan optimization step takes too long
+    *
+    * @param df A dataframe that might require a Catalyst workaround
+    * @return true if a workaround is required
+    */
+  def isCatalystWorkaroundRequired(df: DataFrame, rulesApplied: Int): Boolean = {
+    if (isWorkaroundEnabled) {
+      val elapsedTime = getExecutionPlanGenerationTimeMs(df)
+
+      val tooBigTimeDifference = if (lastExecutionPlanOptimizatioTime > 0) {
+        elapsedTime / lastExecutionPlanOptimizatioTime > 5
+      } else {
+        false
+      }
+      val tooLong = tooBigTimeDifference ||
+        elapsedTime > baselineTimeMs + maxToleratedPlanGenerationPerRuleMs * rulesApplied
+
+      val msg = s"Execution optimization time for $rulesApplied rules: $elapsedTime ms"
+      if (tooLong) {
+        log.warn(s"$msg (Too long!)")
+      } else {
+        baselineTimeMs = Math.max(baselineTimeMs, elapsedTime)
+        lastExecutionPlanOptimizatioTime = elapsedTime
+        log.info(s"New baseline execution plan optimization time: $baselineTimeMs ms")
+        log.warn(msg)
```

Comment:
**Question**: `warn` even in the OK branch? Isn't this more like `debug`?

Reply:
Good spot. But I'd like to leave it like that right now so that integration tests that currently suppress INFO messages still display as much timing information as possible.